src/reformulate.jl

```diff
@@ -1,82 +1,454 @@
+end
+function _reformulate_disjunction(model::JuMP.Model, disj::ConstraintData{T}, method::Hull) where {T<:Disjunction}
+    disj_vrefs = _get_disjunction_variables(model, disj)
+    _update_variable_bounds.(disj_vrefs)
+    hull = _Hull(method.value, disj_vrefs)
+    for d in disj.constraint.disjuncts #reformulate each disjunct
+        _disaggregate_variables(model, d, disj_vrefs, hull) #disaggregate variables for that disjunct
+        _reformulate_disjunct(model, d, hull)
     end
-    #run reformulation
-    if reformulation == :hull
-        disaggregate_variables(m, disj, bin_var)
-        sum_disaggregated_variables(m, disj, bin_var)
+    for vref in disj_vrefs #create sum constraint for disaggregated variables
+        _aggregate_variable(model, vref, hull)
     end
-    reformulate_disjunction(disj, bin_var, reformulation, param)
+end
 
-    #show new constraints as a Dict
-    new_constraints = Dict{Symbol,Any}(
-        Symbol(bin_var,"[$i]") => disj[i] for i in eachindex(disj)
-    )
-    new_constraints[Symbol(bin_var,"_XOR")] = constraint_by_name(m, "XOR(disj_$bin_var)")
-    if reformulation == :hull
-        for var in m[:gdp_variable_refs]
-            agg_con_name = "$(var)_$(bin_var)_aggregation"
-            new_constraints[Symbol(agg_con_name)] = constraint_by_name(m, agg_con_name)
-        end
+# individual disjuncts
+function _reformulate_disjunct(model::JuMP.Model, ind_idx::LogicalVariableIndex, method::AbstractReformulationMethod)
+    #reformulate each constraint and add to the model
+    bv_idx = _indicator_to_binary(model)[ind_idx]
+    bvref = JuMP.VariableRef(model, bv_idx)
+    for cidx in _indicator_to_constraints(model)[ind_idx]
+        cdata = _disjunct_constraints(model)[cidx]
+        reformulate_disjunct_constraint(model, cdata.constraint, bvref, method)
     end
-    return new_constraints
+end
 
-    #remove model.optimize_hook ?
+################################################################################
+#                              BIG-M REFORMULATION
+################################################################################
+function reformulate_disjunct_constraint(
+    model::JuMP.Model,
+    con::JuMP.ScalarConstraint{T, S}, 
+    bvref::JuMP.VariableRef,
+    method::BigM
+) where {T, S <: _MOI.LessThan}
+    M = _get_M_value(method, con.func, con.set)
+    new_func = JuMP.@expression(model, con.func - M*(1-bvref))
+    reform_con = JuMP.add_constraint(model, 
+        JuMP.build_constraint(error, new_func, con.set)
+    )
+    push!(_reformulation_constraints(model), 
+        (JuMP.index(reform_con), JuMP.ScalarShape())
+    )
+end
+function reformulate_disjunct_constraint(
+    model::JuMP.Model,
+    con::JuMP.VectorConstraint{T, S, R}, 
+    bvref::JuMP.VariableRef,
+    method::BigM
+) where {T, S <: _MOI.Nonpositives, R}
+    M = JuMP.@expression(model, [i=1:con.set.dimension],
+        _get_M_value(method, con.func[i], con.set)
+    )
```

Comment:
We should just use a comprehension here instead of `@expression` since this is just making a vecotr of constants.

Reply:
Why is that preferred?

Reply:
Well, we aren't generating any expressions here. So all the extra overhead from `@expression` is not needed and at first glance it seems to imply `M` will be a vector of JuMP expressions. More compactly we can just write:
```julia
M = [_get_M_value(method, con.func[i], con.set) for i in 1:con.set.dimension]
```

Reply:
done